cpp/src/session.cpp

```diff
@@ -148,4 +158,209 @@ void* session::user_data() const {
     return sctx.user_data_;
 }
 
+class transaction_impl {
+  public:
+    proton::sender txn_ctrl;
+    proton::messaging_handler *handler = nullptr;
+    proton::binary transaction_id;
+    bool failed = false;
+    enum State {
+      FREE,
+      DECLARING,
+      DECLARED,
+      DISCHARGING,
+    };
+    enum State state = State::FREE;
+    std::vector<proton::tracker> pending;
+
+    void commit();
+    void abort();
+    void declare();
+
+    void discharge(bool failed);
```

Comment:
For example I'm pretty sure the `discharge()` is only used internal to here so should be private.